ratis-server/src/main/java/org/apache/ratis/server/impl/ServerState.java

```diff
@@ -110,22 +111,31 @@
         RaftServerConfigKeys.Log.corruptionPolicy(prop));
     snapshotManager = new SnapshotManager(storage, id);
 
-    long lastApplied = initStatemachine(stateMachine, group.getGroupId());
+    initStatemachine(stateMachine, group.getGroupId());
 
     // On start the leader is null, start the clock now
     leaderId = null;
     this.lastNoLeaderTime = Timestamp.currentTime();
     this.noLeaderTimeout = RaftServerConfigKeys.Notification.noLeaderTimeout(prop);
 
+    LongSupplier snapshotIndexSupplier = () -> {
+      SnapshotInfo snapshot = stateMachine.getLatestSnapshot();
+      if (snapshot == null || snapshot.getTermIndex().getIndex() < 0) {
+        return RaftLog.INVALID_LOG_INDEX;
+      }
+      return snapshot.getIndex();
+    };
+
     // we cannot apply log entries to the state machine in this step, since we
     // do not know whether the local log entries have been committed.
-    this.log = initRaftLog(getMemberId(), server, storage, this::setRaftConf, lastApplied, prop);
+    this.log = initRaftLog(getMemberId(), server, storage, this::setRaftConf, snapshotIndexSupplier, prop);
 
     RaftLog.Metadata metadata = log.loadMetadata();
     currentTerm.set(metadata.getTerm());
     votedFor = metadata.getVotedFor();
 
-    stateMachineUpdater = new StateMachineUpdater(stateMachine, server, this, lastApplied, prop);
+    stateMachineUpdater = new StateMachineUpdater(stateMachine, server, this,
+        snapshotIndexSupplier.getAsLong(), prop);
```

Comment:
snapshotIndexSupplier.getAsLong() is implemented by StateMachine so that it could be expensive.  Let's use log.getSnapshotIndex().
```
    stateMachineUpdater = new StateMachineUpdater(stateMachine, server, this, log.getSnapshotIndex(), prop);
```